systest/tests/distributed_post_verification_test.go

```diff
@@ -0,0 +1,312 @@
+	)
+	require.NoError(t, err)
+
+	grpcPostService := grpcserver.NewPostService(logger.Named("grpc-post-service"))
+	grpczap.SetGrpcLoggerV2(grpclog, logger.Named("grpc"))
+	grpcPrivateServer, err := grpcserver.NewWithServices(
+		cfg.API.PostListener,
+		logger.Named("grpc-server"),
+		cfg.API,
+		[]grpcserver.ServiceAPI{grpcPostService},
+	)
+	require.NoError(t, err)
+	require.NoError(t, grpcPrivateServer.Start())
+	t.Cleanup(func() { assert.NoError(t, grpcPrivateServer.Close()) })
+
+	nipostBuilder, err := activation.NewNIPostBuilder(
+		localsql.InMemory(),
+		activation.NewPoetDb(sql.InMemory(), log.NewNop()),
+		grpcPostService,
+		cfg.PoetServers,
+		logger.Named("nipostBuilder"),
+		signer,
+		cfg.POET,
+		clock,
+	)
+	require.NoError(t, err)
+
+	// 2.1. Create initial POST
+	var challenge *types.NIPostChallenge
+	for {
+		client, err := grpcPostService.Client(signer.NodeID())
+		if err != nil {
+			ctx.Log.Info("waiting for poet service to connect")
+			time.Sleep(time.Second)
+			continue
+		}
+		ctx.Log.Info("poet service to connected")
+		post, postInfo, err := client.Proof(ctx, shared.ZeroChallenge)
+		require.NoError(t, err)
+
+		challenge = &types.NIPostChallenge{
+			PrevATXID:      types.EmptyATXID,
+			PublishEpoch:   2,
+			PositioningATX: goldenATXID,
+			CommitmentATX:  &postInfo.CommitmentATX,
+			InitialPost:    post,
+		}
+		break
+	}
+
+	nipost, err := nipostBuilder.BuildNIPost(ctx, challenge)
+	require.NoError(t, err)
+
+	// 2.2 Create ATX with invalid POST
+	for i := range nipost.Post.Indices {
+		nipost.Post.Indices[i] += 1
+	}
+	// Sanity check that the POST is invalid
+	verifier, err := activation.NewPostVerifier(cfg.POST, logger.Named("post-verifier"))
+	require.NoError(t, err)
+	err = verifier.Verify(ctx, (*shared.Proof)(nipost.Post), &shared.ProofMetadata{
+		NodeId:          signer.NodeID().Bytes(),
+		CommitmentAtxId: challenge.CommitmentATX.Bytes(),
+		NumUnits:        nipost.NumUnits,
+		Challenge:       nipost.PostMetadata.Challenge,
+		LabelsPerUnit:   nipost.PostMetadata.LabelsPerUnit,
+	})
+	var invalidIdxError *verifying.ErrInvalidIndex
+	require.ErrorAs(t, err, &invalidIdxError)
+
+	atx := types.NewActivationTx(
+		*challenge,
+		types.Address{1, 2, 3, 4},
+		nipost.NIPost,
+		nipost.NumUnits,
+		&nipost.VRFNonce,
+	)
+	nodeID := signer.NodeID()
+	atx.InnerActivationTx.NodeID = &nodeID
+	require.NoError(t, activation.SignAndFinalizeAtx(signer, atx))
+
+	// 3. Wait for publish epoch
+	epoch := atx.PublishEpoch
+	logger.Sugar().Infow("waiting for publish epoch", "epoch", epoch, "layer", epoch.FirstLayer())
+	err = layersStream(ctx, cl.Client(0), logger, func(resp *pb.LayerStreamResponse) (bool, error) {
+		logger.Info("new layer", zap.Uint32("layer", resp.Layer.Number.Number))
+		return resp.Layer.Number.Number < epoch.FirstLayer().Uint32(), nil
+	})
+	require.NoError(t, err)
+
+	// 4. Publish ATX
+	publishCtx, stopPublishing := context.WithCancel(ctx.Context)
+	t.Cleanup(stopPublishing)
```

Comment:
NIT: generally you want to call `context.CancelFunc`s at the end of the test, not during cleanup:

```suggestion
	defer stopPublishing()
```

These are executed before `t.Cleanup` so in the cleanup step you know that all requests have been canceled already.